File: python/046. Permutations.py
class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        length = len(nums)
        if length == 0: return []
        if length == 1: return [nums]
        res = []
        for i in range(length):
            for combine in self.permute(nums[:i] + nums[i+1:]):
                res.append([nums[i]] + combine)
        return res

File: C++/046. Permutations.cpp
class Solution {
    void searchPermute(vector<int>& nums, unordered_map<int,int> &hash, int count, vector<int> list, vector<vector<int>> &res)
    {
        if(count==0){res.push_back(list);return;}
        for(int i=0;i<nums.size();i++)
        {
            if(!hash[nums[i]])
            {
                list.push_back(nums[i]);
                hash[nums[i]]=1;
                searchPermute(nums, hash, count-1, list, res);
                hash[nums[i]]=0;
                list.pop_back();
            }
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> res={};
        unordered_map<int,int> hash;
        vector<int> list={};
        searchPermute(nums, hash, n, list, res);
        return res;
    }
};
